### PR TITLE
DOC: add a "big picture" roadmap

### PR DESCRIPTION
**Pull request by rgommers**
Follow-up to gh-9489, it splits things into a big picture roadmap (idea: 5-10 of our top prios) and a detailed roadmap (which is what we had before).

From the seven ideas currently listed on the "big picture" roadmap, only *Support for distributed arrays and GPU arrays* is new. The other ones are all present on the detailed roadmap or issues. This is my impression of what the highest prios are. Additions, criticism, questions etc. all very welcome of course, I'm sure we can improve on this.

### CODE REVIEW

File: doc/source/roadmap.rst [hunk 1]
```diff
@@ -1 +1,104 @@
...
+is in good shape, however we can make a number of improvements:
+
+1. Library support. Our released wheels now ship with OpenBLAS, which is
+currently the only feasible performant option (ATLAS is too slow, MKL cannot be
+the default due to licensing issues, Accelerate support is dropped because
+Apple doesn't update Accelerate anymore). OpenBLAS isn't very stable though,
+sometimes its releases break things and it has issues with threading (currently
+the only issue for using SciPy with PyPy3).  We need at the very least better
+support for debugging OpenBLAS issues, and better documentation on how to build
+SciPy with it.  And option is to use BLIS for a BLAS interface (see `numpy
```
Comment (tylerjereddy) [inline doc/source/roadmap.rst]:
```suggestion
SciPy with it.  An option is to use BLIS for a BLAS interface (see `numpy
```

File: doc/source/roadmap.rst [hunk 2]
```diff
@@ -1 +1,104 @@
...
+parts of the SciPy API work, and improve support over time.
+
+
+Improve benchmark system for ``optimize``
+-----------------------------------------
+
+``scipy.optimize`` has an extensive set of benchmarks for accuracy and speed of
+the global optimizers. That has allowed adding new optimizers (``shgo`` and
+``dual_annealing``) with significantly better performance than the existing
+ones.  The benchmark system itself is slow and hard to use however; we need to
```
Comment (tylerjereddy) [inline doc/source/roadmap.rst]: `The benchmark system itself` here means `asv` or something else?

File: doc/source/roadmap.rst [hunk 3]
```diff
@@ -1 +1,104 @@
...
+Recently all known issues with ``optimize.linprog`` have been solved.  Now we
+have many ideas for additional functionality (e.g. integer constraints, sparse
+matrix support, performance improvements), see gh-9269.
+
+
+Improve source builds on Windows
+--------------------------------
+
+SciPy critically relies on Fortran code. This is still problematic on Windows.
+There are currently only two options: using Intel Fortran), or using
```
Comment (tylerjereddy) [inline doc/source/roadmap.rst]: spurious bracket here I think

File: doc/source/roadmap.rst [hunk 4]
```diff
@@ -1 +1,104 @@
...
+performance improvements (see `this NumPy PR
+<https://github.com/numpy/numpy/pull/11888>`__ for details),
+add a backend system to support PyFFTW and mkl-fft,
+and align the function signatures of ``numpy.fft`` and ``scipy.fftpack``.
+
+
+Support for distributed arrays and GPU arrays
+---------------------------------------------
+
+NumPy is splitting its API from its execution engine with
```
Comment (tylerjereddy) [inline doc/source/roadmap.rst]: Nice -- very apt description of these dispatcher functionalities.

File: doc/source/roadmap.rst [hunk 5]
```diff
@@ -1 +1,104 @@
...
+
+Improve source builds on Windows
+--------------------------------
+
+SciPy critically relies on Fortran code. This is still problematic on Windows.
+There are currently only two options: using Intel Fortran), or using
+MSVC + gfortran.  The former is expensive, while the latter works (it's what we
+use for releases) but is quite hard to do correctly.  For allowing contributors
+and end users to reliably build SciPy on Windows, using the Flang compiler
+looks like the best way forward long-term.  Until Flang support materializes,
```
Comment (tylerjereddy) [inline doc/source/roadmap.rst]: Cool, I'm not familiar with `Flang` but definitely agree this stuff is painful as I've worked through the Azure stuff for SciPy  windows builds (64-bit builds just starting to work after what seemed like days of guesswork!).

File: doc/source/roadmap.rst (hunk 2 above)
Comment (rgommers) [reply doc/source/roadmap.rst]: Specifically the `optimize` benchmarks. You need to set `SCIPY_XSLOW`, use `SCIPY_GLOBAL_BENCH`, and then it takes forever. Plus there's no good way to compare outputs. For `shgo` and `dual_annealing` both authors forked the benchmarks so they could actually work with them.

I'll add `optimize` to "the benchmark system" to clarify.

Comment (rgommers) [reply doc/source/roadmap.rst]: I'm actually not 100% sure that this global benchmarks idea needs to be here - probably the least important of everything I put in this list.

Comment (andyfaff) [reply doc/source/roadmap.rst]: I may be biased but I wouldn't say that the shgo and dual_annealing are *significantly* better than basinhopping or differential_evolution. THey're all pretty good.

Comment (andyfaff) [reply doc/source/roadmap.rst]: Yeah, the global benchmarks are not straightforward to use, could definitely do with more documentation.

Comment (larsoner) [reply doc/source/roadmap.rst]: It would be great if we could actually get #8779 working. It could serve as a way to show people how to run the benchmarks.

File: doc/source/roadmap-detailed.rst [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+We probably want to deprecate fftpack.convolve as public function (it was not
+meant to be public).
+
+There's a large overlap with ``numpy.fft``.  This duplication has to change
+(both are too widely used to deprecate one); in the documentation we should
+make clear that ``scipy.fftpack`` is preferred over ``numpy.fft``.
+If there are differences in signature or functionality, the best version
+should be picked case by case (example: numpy's ``rfft`` is preferred, see
+gh-2487).
```
Comment (larsoner) [inline doc/source/roadmap-detailed.rst]:
```
`gh-2487 <https://github.com/scipy/scipy/issues/2487>`__
```

File: doc/source/roadmap-detailed.rst [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+integrate
+`````````
+Needed for ODE solvers:
+
+  - Documentation is pretty bad, needs fixing
+  - A new ODE solver interface  (``solve_ivp``) was added in SciPy 1.0.0.
+    In the future we can consider (soft-)deprecating the older API.
+
+The numerical integration functions are in good shape.  Support for integrating
+complex-valued functions and integrating multiple intervals (see gh-3325) could
```
Comment (larsoner) [inline doc/source/roadmap-detailed.rst]:
```
`gh-3325 <https://github.com/scipy/scipy/issues/3325>`__
```

File: doc/source/roadmap-detailed.rst [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+
+optimize
+````````
+Overall this module is in good shape. Two good global optimizers were added in
+1.2.0; large-scale optimizers is still a gap that could be filled.  Other
+things that are needed:
+
+  - Many ideas for additional functionality (e.g. integer constraints, sparse
+    matrix support, performance improvements) in ``linprog``, see gh-9269.
```
Comment (larsoner) [inline doc/source/roadmap-detailed.rst]:
```
`gh-9269 <https://github.com/scipy/scipy/issues/9269>`__
```

File: doc/source/roadmap-detailed.rst [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+
+    1. Get good double-precision implementations. This is doable for parabolic
+       cylinder functions (in progress). I think it's possible for hypergeometric
+       functions, though maybe not in time. For spheroidal wavefunctions this is
+       not possible with current theory.
+
+    2. Port Boost's arbitrary precision library and use it under the hood to get
+       double precision accuracy. This might be necessary as a stopgap measure
+       for hypergeometric functions; the idea of using arbitrary precision has
+       been suggested before by @nmayorov and in gh-5349. Likely necessary for
```
Comment (larsoner) [inline doc/source/roadmap-detailed.rst]:
```
`gh-5349 <https://github.com/scipy/scipy/issues/5349>`__
```

File: doc/source/roadmap.rst [hunk 10]
```diff
@@ -1 +1,104 @@
...
+
+1. Library support. Our released wheels now ship with OpenBLAS, which is
+currently the only feasible performant option (ATLAS is too slow, MKL cannot be
+the default due to licensing issues, Accelerate support is dropped because
+Apple doesn't update Accelerate anymore). OpenBLAS isn't very stable though,
+sometimes its releases break things and it has issues with threading (currently
+the only issue for using SciPy with PyPy3).  We need at the very least better
+support for debugging OpenBLAS issues, and better documentation on how to build
+SciPy with it.  An option is to use BLIS for a BLAS interface (see `numpy
+gh-7273 <https://github.com/numpy/numpy/issues/7372>`__).
```
Comment (larsoner) [inline doc/source/roadmap.rst]: says 7273 but links to 7372 (the 7273 appears to be the typo)

**Review by larsoner — COMMENTED** [review]
Looks good to me other than a few minor nitpicks/suggestions.

In theory a lot of the double-backtick namings could be converted to the form
```
:mod:`scipy.signal`
```
and the like, but that can be done in another PR when someone is bored. Unless @rgommers you are already looking for something more to do :)

File: doc/source/roadmap-detailed.rst [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+linalg
+``````
+``scipy.linalg`` is in good shape.  We have started requiring more recent
+LAPACK versions (minimum version increases from 3.1.0 to 3.4.0 in SciPy 1.2.0);
+we want to add support for newer features in LAPACK.
+
+Needed:
+
+  - Reduce duplication of functions with ``numpy.linalg``, make APIs consistent.
+  - ``get_lapack_funcs`` should always use ``flapack``
```
Comment (ilayn) [inline doc/source/roadmap-detailed.rst]: I can do this right away in #9455, is there any specific thing to do?

File: doc/source/roadmap-detailed.rst [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+
+linalg
+``````
+``scipy.linalg`` is in good shape.  We have started requiring more recent
+LAPACK versions (minimum version increases from 3.1.0 to 3.4.0 in SciPy 1.2.0);
+we want to add support for newer features in LAPACK.
+
+Needed:
+
+  - Reduce duplication of functions with ``numpy.linalg``, make APIs consistent.
```
Comment (ilayn) [inline doc/source/roadmap-detailed.rst]: where can we start this discussion? Mailing list?

Comment (rgommers) [reply doc/source/roadmap-detailed.rst]: I think it's mostly about "making APIs consistent". We can't remove much, there will be duplicate functions. I'd say the important thing is to have an overview of where APIs don't match, and then figure out which one is better. One example that is discussed in an issue somewhere: the modes of `qr` don't (or didn't?) match. 

Note, I just moved this content, didn't think about it recently.

File: doc/source/roadmap-detailed.rst (hunk 11 above)
Comment (rgommers) [reply doc/source/roadmap-detailed.rst]: No memory of this one. Maybe it already does the right thing? This has been on here for some years.

Comment (ilayn) [reply doc/source/roadmap-detailed.rst]: Indeed it is quite trivial to fix but I don't know if there was any specific detail about it. @larsoner @ev-br @pv Anybody remembers the reason for this? 

We can simply get rid of this switch below and we are done as far as I can see.

https://github.com/scipy/scipy/blob/master/scipy/linalg/blas.py#L318

Comment (ev-br) [reply doc/source/roadmap-detailed.rst]: This is before my time, sorry. 
Anyway, if it felt desirable in 2013 already, if it simplifies maintenance and does not break the world, why not. If done, this should be done in a small self-contained PR early in the release cycle, so that downstream has time to complain (or, possibly, carry a patch if they do).

File: doc/source/roadmap.rst (hunk 2 above)
Comment (rlucas7) [reply doc/source/roadmap.rst]: > It would be great if we could actually get #8779 working. It could serve as a way to show people how to run the benchmarks.

+1 I would love to know how to run the benchmarks

**Comment by chrisb83** [comment]
I really like the idea of having a detailed roadmap. Would this also be supposed to help potential contributors who are interested in adding features but who don't know where to start?I think it would be nice to discuss some ideas for the different modules to get a better picture of the priorities (not sure if here or in separate issues per module?). 

A couple of questions / ideas came to my mind for `scipy.stats`:

- If we want to add general ideas for contributors: implementing (well-kown) distributions is always welcome I guess.
- What about issues that are considered high priority? (Maybe issues related to mannwithneyu?. This would imply regular maintenance of the roadmap toegther with a new release)
- what about mstats (in particular, documentation / consistency with stats). Warren's comment in #9395 just reminded me about this.

**Comment by rlucas7** [comment]
I agree with @chrisb83 's comments on detailed roadmaps for sub-modules. Having a good idea of priorities and more specific guidance on where to start would be helpful for those who haven't contributed before and for those (like me) who've done a couple commits but lack the knowledge and experience of the members.

**Comment by rgommers** [comment]
> I really like the idea of having a detailed roadmap. Would this also be supposed to help potential contributors who are interested in adding features but who don't know where to start?I think it would be nice to discuss some ideas for the different modules to get a better picture of the priorities (not sure if here or in separate issues per module?).

Yes, that was indeed one of the main ideas for the roadmap that we've had for a while. As long as we don't make it significantly more detailed than what's there now (we can use issues labelled `enhancement` for that) then having a clear list of prios per submodule would be great.


>    If we want to add general ideas for contributors: implementing (well-kown) distributions is always welcome I guess.

+1 

> What about issues that are considered high priority? (Maybe issues related to mannwithneyu?. This would imply regular maintenance of the roadmap toegther with a new release)

I would suggest only if they're structural, not single issues for one function. Or if we do that, maybe super short: "high prio issues are for mannwhitneyu (`gh-xxxx`), ... and ...

> What about mstats (in particular, documentation / consistency with stats). Warren's comment in #9395 just reminded me about this.

That got a lot better over time, but indeed there's still quite some work to do. That fits well on the roadmap I'd say.

**Comment by rgommers** [comment]
If there are more comments on the "big picture" part, by tomorrow would be very useful. Reason: @scopatz is hosting a webinar and he invited me to talk about the SciPy roadmap this Friday ([details](https://www.quansight.com/open-source-directions), [register](https://app.livestorm.co/quansight )). By then we hope to get a brochure out as well, similar to the ones for other projects on  https://www.quansight.com/projects.

To recap, the 7 listed ideas in order of prio are:
- Evolve BLAS and LAPACK support
- Sparse arrays
- Fourier transforms
- Support for distributed arrays and GPU arrays
- Improve benchmark system for optimize
- Linear programming enhancements
- Improve source builds on Windows

**Comment by scopatz** [comment]
Thanks for putting this in @rgommers!

File: doc/source/roadmap.rst (hunk 2 above)
Comment (rgommers) [reply doc/source/roadmap.rst]: > I may be biased but I wouldn't say that the shgo and dual_annealing are significantly better than basinhopping or differential_evolution. 

Based on the recent benchmarks, `basinhopping` < `differential_evolution` < `dual_annealing` / `shgo`. After using the new additions recently, I'm not yet convinced that `dual_annealing` is significantly better than `differential_evolution`, but I'm quite impression by `shgo` (also its ability to find *all* minima is very handy.

But yeah, I remember the pre-`basinhopping` days, compared to back then they're all awesome:)

Comment (rgommers) [reply doc/source/roadmap.rst]: > +1 I would love to know how to run the benchmarks

The basics are straightforward: `python runtests.py --bench`. However, when we get into comparing benchmarks between branches etc., it tends to get fragile.

File: doc/source/roadmap.rst [hunk 13]
```diff
@@ -1 +1,105 @@
...
+Support for distributed arrays and GPU arrays
+---------------------------------------------
+
+NumPy is splitting its API from its execution engine with
+``__array_function__`` and ``__array_ufunc__``.  This will enable parts of SciPy
+to accept distributed arrays (e.g. ``dask.array.Array``) and GPU arrays (e.g.
+``cupy.ndarray``) that implement the ``ndarray`` interface.  At the moment it is
+not yet clear which algorithms will work out of the box, and if there are
+significant performance gains when they do.  We want to create a map of which
+parts of the SciPy API work, and improve support over time.
```
Comment (shoyer) [inline doc/source/roadmap.rst]: In addition to making use of NumPy protocols like `__array_function__`, there's the possibility of supporting these protocols in SciPy as well, e.g., so it's possible to (re)implement scipy functions like those in `scpiy.signal` for dask or gpu arrays:
http://www.numpy.org/neps/nep-0018-array-function-protocol.html#use-outside-of-numpy

Is there interest in exploring this as well?

Comment (rgommers) [reply doc/source/roadmap.rst]: Yes I think so. A bit lower on the prio list probably, simply because I think we'd like to see first how this works out for numpy. But could be quite nice. What do others think?

**Comment by rgommers** [comment]
Any other comments? I'm thinking of merging this at the start of the new year.

File: doc/source/roadmap-detailed.rst [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+<https://github.com/scipy/scipy/issues/3325>`__) could be added.
+
+
+interpolate
+```````````
+
+Ideas for new features:
+
+  - Spline fitting routines with better user control.
+  - Integration and differentiation and arithmetic routines for splines
```
Comment (ev-br) [inline doc/source/roadmap-detailed.rst]: Integration and differentiation are implemented, not sure what is meant by arithmetic routines.

File: doc/source/roadmap-detailed.rst [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+
+stats
+`````
+
+This module is in good shape overall.  New functionality that's similar to
+what's already present can continue to be added; more advanced statistical
+routines may fit better in ``statsmodels``.  Some ideas for new contributions
+are:
+
+- Implementing (well-kown) distributions to the ``stats.distributions``
```
Comment (ev-br) [inline doc/source/roadmap-detailed.rst]: Typo: kown

File: doc/source/roadmap-detailed.rst [hunk 16]
```diff
@@ -0,0 +1,385 @@
...
+
+
+stats
+`````
+
+This module is in good shape overall.  New functionality that's similar to
+what's already present can continue to be added; more advanced statistical
+routines may fit better in ``statsmodels``.  Some ideas for new contributions
+are:
+
```
Comment (ev-br) [inline doc/source/roadmap-detailed.rst]: Maybe also mention Bunches: almost PR-ready implementations are well hidden in (possibly closed )issues concerning linregress.

**Review by ev-br — COMMENTED** [review]
Looks good, just a couple of minor comments.

File: doc/source/roadmap-detailed.rst (hunk 16 above)
Comment (rgommers) [reply doc/source/roadmap-detailed.rst]: good idea, added

File: doc/source/roadmap-detailed.rst (hunk 14 above)
Comment (rgommers) [reply doc/source/roadmap-detailed.rst]: agree, removed

**Comment by chrisb83** [comment]
One more suggestion for `stats`: improve statistical tests (p-value calculation / alternative hypothesis). Here R often has better functionality.

Examples:
- exact two-sided KS test (Marsaglia et al)
- two-sided vs one-sided tests (Wilcoxon)

File: doc/source/roadmap-detailed.rst [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+More importantly, still, SciPy implements one *variant* of this data
+point model, where datapoints at any two extremes of an axis share a
+spatial location under *periodic wrapping* mode.  E.g., in a 1D array,
+you would have ``x[0]`` and ``x[-1]`` co-located.  A very common
+use-case, however, is for signals to be periodic, with equal spacing
+between the first and last element along an axis (instead of zero
+spacing).  Adding wrapping modes for this use-case and disambiguating
+them from the existing implementation should address several issues,
+including #1323, #1903, #2045 and #2640.
```
Comment (charris) [inline doc/source/roadmap-detailed.rst]: The basis for this was done in #8537, but the interpolation routines need updating. I think that should not be too difficult. @jaimefrio Comment?

Comment (charris) [reply doc/source/roadmap-detailed.rst]: The ndimage interpolation might be considered a mix of the two conventions, which it used depends on the spline order. That is, for odd orders the knot points are at the same location as the data points, while for even orders the knot points are midway between the data points. That is convenient, in that the number of needed boundary conditions is always even and can be rationally split between the beginning and end of the data.

I think the distinction between data point models is artificial. For images the data point is simply taken at the center of the pixel, which makes sense as the output is the sum over the pixel. Even in other applications, understanding the result needs some familiarity with the bandwidth of the signal vs the measurement device and sample frequency.

Comment (rgommers) [reply doc/source/roadmap-detailed.rst]: The required new modes were indeed added recently, thanks @charris. 

> I think the distinction between data point models is artificial. 

Not sure I agree with this. Reality is more complex, but it does help to think about these two modes. IIRC we had a mix of both modes even for the default of odd order.

**Comment by rgommers** [comment]
> One more suggestion for stats: improve statistical tests (p-value calculation / alternative hypothesis). Here R often has better functionality.

Thanks Christoph, added.

File: doc/source/roadmap-detailed.rst (hunk 17 above)
Comment (charris) [reply doc/source/roadmap-detailed.rst]: I should probably clarify. We have five types of boundary conditions, the spline coefficients can now be computed for all of them. I believe that interpolation also works for all of them (@jaimefrio ). The "mirror" and "reflect" modes correspond to the boundary conditions for DCT-I and DCT-II. The different boundary conditions all use the same data points and all can be evaluated outside the interval spanned by the data points, but their behavior outside that interval varies depending on the mode. In "reflect" mode, which is generally preferred for numerical reasons among others, the interval [-.5, 0] has content that may be of interest and cannot be obtained by simply mirroring into interior points. I don't regard this primarily a  difference between pixel/data views, but rather as different boundary conditions. The pixel geometry might lead one to prefer the 'reflect' mode, but really, at the edge, the interpolated results are going to be suspect in any case.

Due to the way the splines are implemented, the correspondence between data points and knot points varies as pointed out previously. That may be counter intuitive, but the major effect is that differentiating or integrating the splines does not yield the knot points expected by the current interpolation routines. I don't know if there is a plan to fix that.

**Comment by rgommers** [comment]
> Any other comments? I'm thinking of merging this at the start of the new year.

Okay, in it goes. We can open new issues/PR or discuss on the mailing list for the rest. Except the `ndimage` interpolation modes discussion, let's continue that here. Needs @stefanv's input, we're tweaking his text (what there now is certainly an improvement over the old text though).
